src/tools/generate/generate_table_of_contents.js

```diff
@@ -0,0 +1,81 @@
+const { JSDOM } = require('jsdom');
+
+const generate_table_of_contents = (html) => {
+  const dom = new JSDOM(html);
+  const all_headings = Object.values(
+    dom.window.document.querySelectorAll('h1, h2, h3, h4, h5, h6')
+  );
+  const starting_level = get_level(all_headings[0]);
+  const nested_headings = nest_headings(all_headings, starting_level);
+  const toc = generate_html(nested_headings.children);
+
+  return toc;
+};
+
+const generate_html = (headings) => {
```

Comment:
Because this recursively generates the table of contents, I've kept it in javascript. If we want to move it into the ejs template, we would need to create it as an partial inside the templates directory.

Reply:
Hmmm yeah this is tricky. I think my preference is to have all HTML in the templates. What do you think of generating a JSON object of the TOC and injecting that into the chapter template, similar to the metadata? Then we can run that through a Jinja macro to generate the TOC markup.

Reply:
I've moved this into a separate ejs template, instead of doing it in template literal.